src/topology/fundamental_groupoid.lean

```diff
@@ -0,0 +1,458 @@
+end
+
+@[simp] lemma inv_id (x : X) : (path.id x).inv = path.id x :=
+ext $ λ s, rfl
+
+end path
+
+structure homotopy {x y : X} (f g : path x y) : Type u :=
+(to_fun : set.Icc (0:ℝ) 1 × set.Icc (0:ℝ) 1 → X)
+(continuous : continuous to_fun)
+(initial : ∀ s, to_fun (s, 0) = x)
+(final : ∀ s, to_fun (s, 1) = y)
+(start : ∀ t, to_fun (0, t) = f.1 t)
+(finish : ∀ t, to_fun (1, t) = g.1 t)
+
+namespace homotopy
+
+variables {w x y z : X} {f g h : path x y}
+
+def refl (f : path x y) : homotopy f f :=
+{ to_fun := λ p, f.1 p.2,
+  continuous := f.2.comp continuous_snd,
+  initial := λ s, f.3,
+  final := λ s, f.4,
+  start := λ t, rfl,
+  finish := λ t, rfl }
+
+def symm (F : homotopy f g) : homotopy g f :=
+{ to_fun := λ p, F.1 (⟨1 - p.1.1, sub_nonneg.2 p.1.2.2,
+    sub_le.2 (by rw sub_self; exact p.1.2.1)⟩, p.2),
+  continuous := F.2.comp $ continuous.prod_mk
+    (continuous_induced_rng $ continuous_sub continuous_const $
+      continuous_induced_dom.comp continuous_fst)
+    continuous_snd,
+  initial := λ s, F.3 _,
+  final := λ s, F.4 _,
+  start := λ t, by simp only [set.Icc_zero_val, sub_zero]; convert F.6 t,
+  finish := λ t, by simp only [set.Icc_one_val, sub_self]; convert F.5 t }
+
+noncomputable def trans (F : homotopy f g) (G : homotopy g h) : homotopy f h :=
+{ to_fun := λ p, if p.1.1 ≤ 0.5
+    then F.1 (⟨min (p.1.1 + p.1.1) 1, le_min (add_nonneg p.1.2.1 p.1.2.1) zero_le_one,
+        min_le_right _ _⟩,
+      p.2)
+    else G.1 (⟨max (p.1.1 + p.1.1 - 1) 0, le_max_right _ _,
+        max_le (sub_le_iff_le_add.2 (add_le_add p.1.2.2 p.1.2.2)) zero_le_one⟩,
+      p.2),
+  continuous := continuous_if
+    (λ p hp, have p.1.1 = 0.5,
+      from frontier_le_subset_eq (continuous_induced_dom.comp continuous_fst) continuous_const hp,
+      by simp only [this, add_halves, sub_self, min_self, max_self]; erw [F.6, G.5])
+    (F.2.comp $ continuous.prod_mk (continuous_induced_rng $
+        continuous_min (continuous_add (continuous_induced_dom.comp continuous_fst)
+            (continuous_induced_dom.comp continuous_fst))
+          continuous_const)
+      continuous_snd)
+    (G.2.comp $ continuous.prod_mk (continuous_induced_rng $
+        continuous_max (continuous_sub (continuous_add (continuous_induced_dom.comp continuous_fst)
+            (continuous_induced_dom.comp continuous_fst))
+          continuous_const) continuous_const)
+      continuous_snd),
+  initial := λ s, by dsimp only; split_ifs; simp only [F.3, G.3],
+  final := λ s, by dsimp only; split_ifs; simp only [F.4, G.4],
+  start := λ t, by erw [if_pos (le_of_lt $ half_pos (zero_lt_one : (0:ℝ)<1))];
+    simp only [set.Icc_zero_val, add_zero, min_eq_left (zero_le_one : (0:ℝ)≤1)]; exact F.5 t,
+  finish := λ t, by erw [if_neg (not_le_of_lt $ half_lt_self (zero_lt_one : (0:ℝ)<1))];
+    simp only [set.Icc_one_val, add_sub_cancel, max_eq_left (zero_le_one : (0:ℝ)≤1)]; exact G.6 t }
+
+noncomputable def comp {α β : path y z} (F : homotopy f g) (G : homotopy α β) :
```

Comment:
Please, don't use α and β for paths. I know it would be an excellent mathematical notation, but in a Lean context these have to be types, and nothing else.

Reply:
These names are confusing anyways because α and β are the same sort of thing (paths) as f and g. I suggest a convention of using names like f₀ and f₁ for paths (or maps or whatever) related by a homotopy; it makes the structure clearer and saves on letters too.